indico/modules/events/persons/client/js/Email.jsx

```diff
@@ -0,0 +1,320 @@
+import TextEditor, {FinalTextEditor} from 'indico/react/components/TextEditor';
+import {FinalCheckbox, FinalDropdown, FinalInput, handleSubmitError} from 'indico/react/forms';
+import {FinalModalForm} from 'indico/react/forms/final-form';
+import {useIndicoAxios} from 'indico/react/hooks';
+import {Param, Plural, PluralTranslate, Singular, Translate} from 'indico/react/i18n';
+import {indicoAxios} from 'indico/utils/axios';
+import {snakifyKeys} from 'indico/utils/case';
+
+import './Email.module.scss';
+
+const getIds = selector =>
+  Array.from(document.querySelectorAll(selector))
+    .filter(e => e.offsetWidth > 0 || e.offsetHeight > 0)
+    .map(e => +e.value);
+
+export function EmailButton({
+  eventId,
+  roleId,
+  personSelector,
+  userSelector,
+  triggerSelector,
+  noAccount,
+  notInvitedOnly,
+}) {
+  const [open, setOpen] = useState(false);
+  const personIds = getIds(personSelector);
+  const userIds = getIds(userSelector);
+
+  useEffect(() => {
+    if (!triggerSelector) {
+      return;
+    }
+    const handler = () => setOpen(true);
+    const element = document.querySelector(triggerSelector);
+    element.addEventListener('click', handler);
+    return () => element.removeEventListener('click', handler);
+  }, [triggerSelector]);
+
+  return (
+    <>
+      {!triggerSelector && (
+        <Translate as={Button} onClick={() => setOpen(true)}>
+          Send email
+        </Translate>
+      )}
+      {open && (
+        <EmailForm
+          eventId={eventId}
+          personIds={personIds}
+          userIds={userIds}
+          roleIds={roleId && [roleId]}
+          noAccount={noAccount}
+          notInvitedOnly={notInvitedOnly}
+          onClose={() => setOpen(false)}
+        />
+      )}
+    </>
+  );
+}
+
+EmailButton.propTypes = {
+  eventId: PropTypes.number.isRequired,
+  roleId: PropTypes.number,
+  personSelector: PropTypes.string,
+  userSelector: PropTypes.string,
+  triggerSelector: PropTypes.string,
+  noAccount: PropTypes.bool,
+  notInvitedOnly: PropTypes.bool,
+};
+
+EmailButton.defaultProps = {
+  roleId: undefined,
+  personSelector: undefined,
+  userSelector: undefined,
+  triggerSelector: undefined,
+  noAccount: false,
+  notInvitedOnly: false,
+};
+
+export function EmailForm({
+  eventId,
+  personIds,
+  roleIds,
+  userIds,
+  onClose,
+  noAccount,
+  notInvitedOnly,
+}) {
+  const [preview, setPreview] = useState(null);
+  const recipientData = {
+    personId: personIds,
+    roleId: roleIds,
+    userId: userIds,
+    noAccount,
+    notInvitedOnly,
+  };
+  const {data, loading} = useIndicoAxios({
+    url: emailMetadataURL({event_id: eventId}),
+    method: 'POST',
+    data: snakifyKeys(recipientData),
+  });
+  const {
+    senders = [],
+    recipients = [],
+    subject: defaultSubject,
+    body: defaultBody,
+    placeholders = [],
+  } = data || {};
+  const count = Object.values(recipientData).reduce((acc, v) => acc + v.length, 0);
```

Comment:
```suggestion
  const count = recipients.length;
```

much less complicated and actually works even when there's extra stuff in recipientData ;)